File: battleship/block.py
import copy

class block:
    def __init__(self, r, c):
        self.r = r
        self.c = c
        self.is_occupied = False
        self.is_hit = False

    def offset_by(self, r, c):
        return block(self.r + r, self.c + c)

    def with_in(self, b1, b2):
        return b1.r <= self.r <= b2.r and b1.c <= self.c <= b2.c

    def cp(self):
        return copy.deepcopy(self)

    def hide_cp(self):
        c = self.cp()
        c.is_occupied = None
        return c

File: battleship/block.rb
module Battleship

  class Block
    attr_reader :r, :c
    attr_accessor :is_occupied, :is_hit

    def initialize(r, c)
      @r, @c = r, c
      @is_occupied = false
      @is_hit = false
    end

    def offset_by(r, c)
      return Block.new(@r + r, @c + c)
    end

    def with_in?(b1, b2)
      @r.between?(b1.r, b2.r) and @c.between?(b1.c, b2.c)
    end

    def hide_cp
      c = self.dup
      c.is_occupied = nil
      return c
    end
  end

end
